src/myrtlespeech/model/hard_lstm.py

```diff
@@ -0,0 +1,419 @@
+        initialisation.
+
+        Args:
+            x: A Tuple where the first element is the rnn sequence input and
+                the second represents the length of these *input* sequences.
+
+            hx: The Optional lstm hidden state: a Tuple[Tensor, Tensor].
+
+        Returns:
+            A Tuple of Tuples; ``res = (a, b), (c, d)`` where
+                ``a`` is the lstm sequence output, ``b`` are the lengths of
+                these output sequences and ``(c, d)`` are the hidden and cell
+                states of the lstm.
+        """
+        inp, lengths = x
+        if hx is None:
+            hx = init_hidden_state(
+                batch=len(lengths),
+                dtype=inp.dtype,
+                hidden_size=self.hidden_size,
+                num_layers=self.rnn.num_layers,
+                bidirectional=self.bidirectional,
+                rnn_type=self.rnn_type,
+            )
+        if self.use_cuda:
+            inp = inp.cuda()
+            if self.rnn_type == RNNType.LSTM:
+                hx = hx[0].cuda(), hx[1].cuda()  # type: ignore
+            else:
+                hx = hx.cuda()  # type: ignore
+
+        y, hid = self.rnn(inp, hx)
+        return (y, lengths), hid
+
+
+def gen_hard_lstm(
+    input_size: int,
+    hidden_size: int,
+    num_layers: int = 1,
+    bias: int = True,
+    dropout: float = 0.0,
+    bidirectional: bool = False,
+    forget_gate_bias: Optional[float] = None,
+    batch_first: bool = False,
+) -> torch.nn.Module:
+    """Hard LSTM constructor.
+
+    Args:
+        See :py:class:`HardLSTM`.
+    """
+    assert not dropout, "Dropout for HardLSTMs is not supported."
+    if bidirectional:
+        layer_type = HardLSTMBidirLayer
+    else:
+        layer_type = HardLSTMLayer
+
+    return StackedLSTM(
+        num_layers,
+        layer_type,
+        input_size=input_size,
+        hidden_size=hidden_size,
+        forget_gate_bias=forget_gate_bias,
+        batch_first=batch_first,
+        bidirectional=bidirectional,
+        bias=bias,
+    )
+
+
+class StackedLSTM(torch.nn.Module):
+    """Multi-layer LSTM.
+
+    Args:
+        num_layers: See :py:class:`HardLSTM`.
+
+        layer_type: The layer type used in the :py:class:`StackedLSTM`.
+
+        input_size: See :py:class:`HardLSTM`.
+
+        hidden_size: See :py:class:`HardLSTM`.
+
+        bias: See :py:class:`HardLSTM`.
+
+        dropout: See :py:class:`HardLSTM`.
+
+        bidirectional: See :py:class:`HardLSTM`.
+
+        forget_gate_bias: See :py:class:`HardLSTM`.
+
+        batch_first: See :py:class:`HardLSTM`.
+    """
+
+    def __init__(
+        self,
+        num_layers: int,
+        layer_type: torch.nn.Module,
+        input_size: int,
+        hidden_size: int,
+        bias: int = True,
+        dropout: float = 0.0,
+        bidirectional: bool = False,
+        forget_gate_bias: Optional[float] = None,
+        batch_first: bool = False,
```

Comment:
Nit: Order these the same as PyTorch's LSTM, ensure the Args and actual parameter ordering is the same and tack on any additional arguments at the end:

```suggestion
        input_size: See :py:class:`HardLSTM`.

        hidden_size: See :py:class:`HardLSTM`.
        
        num_layers: See :py:class:`HardLSTM`.

        bias: See :py:class:`HardLSTM`.

        batch_first: See :py:class:`HardLSTM`.

        dropout: See :py:class:`HardLSTM`.

        bidirectional: See :py:class:`HardLSTM`.

        forget_gate_bias: See :py:class:`HardLSTM`.
        
        layer_type: The layer type used in the :py:class:`StackedLSTM`.
    """

    def __init__(
        self,
        input_size: int,
        hidden_size: int,
        num_layers: int,
        bias: int = True,
        batch_first: bool = False,
        dropout: float = 0.0,
        bidirectional: bool = False,
        forget_gate_bias: Optional[float] = None,
        layer_type: torch.nn.Module,
```

Reply:
Done!